From: Etienne Millon <me@emillon.org>
Date: Sun, 15 Nov 2015 18:46:28 +0100
Forwarded: https://github.com/sahib/glyr/pull/67
Subject: [PATCH] Use BUILD_DATE / BUILD_TIME instead of cpp macros

`glyrc --version` returns the date and time the library was built. To do that,
it uses the standard macros `__DATE__` and `__TIME__`.

The problem with this is that it makes the build not reproducible, since a
binary produced a few seconds later will obviously be different.

This adds a mechanism to bypass this: if the environment variables `BUILD_DATE`
and `BUILD_TIME` are exported at compile-time, they will be used instead.

This is part of an effort to make Debian packages reproducible, but should
benefit other distributions. More information can be found at [1], and [2] for
this particular issue.

https://wiki.debian.org/ReproducibleBuilds/About
https://wiki.debian.org/ReproducibleBuilds/TimestampsFromCPPMacros
---
 lib/CMakeLists.txt | 6 ++++++
 lib/glyr.c         | 9 ++++++++-
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index 528f32e..a350440 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -6,6 +6,12 @@ configure_file (
 
 SET(GENERIC_LIB_VERSION ${GLYR_VERSION_MAJOR}.${GLYR_VERSION_MINOR})
 SET(GLYR_API_SOVERSION 1)
+IF(DEFINED ENV{BUILD_DATE})
+    ADD_DEFINITIONS(-DBUILD_DATE="$ENV{BUILD_DATE}")
+ENDIF()
+IF(DEFINED ENV{BUILD_TIME})
+    ADD_DEFINITIONS(-DBUILD_TIME="$ENV{BUILD_TIME}")
+ENDIF()
 
 # Link libglyr as shared library
 ADD_LIBRARY(glyr SHARED	${LIB_SOURCE_LOCATIONS})
diff --git a/lib/glyr.c b/lib/glyr.c
index aad2ce5..a1c2ce6 100644
--- a/lib/glyr.c
+++ b/lib/glyr.c
@@ -243,10 +243,17 @@ GlyrMemCache * glyr_cache_copy (GlyrMemCache * cache)
 
 /////////////////////////////////
 
+#ifndef BUILD_DATE
+#define BUILD_DATE __DATE__
+#endif
+#ifndef BUILD_TIME
+#define BUILD_TIME __TIME__
+#endif
+
 __attribute__ ( (visibility ("default") ) )
 const char * glyr_version (void)
 {
-    return "Version "GLYR_VERSION_MAJOR"."GLYR_VERSION_MINOR"."GLYR_VERSION_MICRO" ("GLYR_VERSION_NAME") of ["__DATE__"] compiled at ["__TIME__"]";
+    return "Version "GLYR_VERSION_MAJOR"."GLYR_VERSION_MINOR"."GLYR_VERSION_MICRO" ("GLYR_VERSION_NAME") of ["BUILD_DATE"] compiled at ["BUILD_TIME"]";
 }
 
 /////////////////////////////////
-- 
2.1.4

